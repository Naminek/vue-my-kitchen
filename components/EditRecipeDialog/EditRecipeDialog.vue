<template>
  <v-dialog
    :value="value"
    max-width="1000px"
    scrollable
    persistent
    @input="close"
  >
    <v-card>
      <v-card-title class="blue-grey lighten-2 white--text font-weight-light text-subtitle-1 text-sm-h6">
        Add recipe
        <v-spacer />
        <v-btn
          icon
          class="white--text ml-2"
          @click="close"
        >
          <v-icon>{{ mdiClose }}</v-icon>
        </v-btn>
      </v-card-title>
      <v-card-text class="pt-4">
        <v-form ref="recipeForm" v-model="validates" lazy-validation>
          <RecipeForm :recipe="recipeBuffer" />
        </v-form>
      </v-card-text>
      <v-divider />
      <v-card-actions>
        <v-btn
          text
          class="mx-2 mx-sm-0"
          :icon="$vuetify.breakpoint.xsOnly"
          :disabled="!validates"
          @click="$emit('clicked-save')"
          data-cy="saveButton"
        >
          <v-icon :left="!!$vuetify.breakpoint.smAndUp">
            {{ mdiCheck }}
          </v-icon>
          <span v-if="!!$vuetify.breakpoint.smAndUp">Save</span>
        </v-btn>
        <v-btn
          text
          :icon="$vuetify.breakpoint.xsOnly"
          @click="$emit('clicked-cancel')"
          data-cy="cancelButton"
        >
          <v-icon :left="!$vuetify.breakpoint.smAndUp">
            {{ mdiCancel }}
          </v-icon>
          <span v-if="!!$vuetify.breakpoint.smAndUp">Cancel</span>
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import {
  PropType, defineComponent, ref, watch, toRef
} from '@nuxtjs/composition-api'
import { mdiCancel, mdiCheck, mdiClose } from '@mdi/js'
import Recipe from '~/models/recipe'
import RecipeForm from '~/components/RecipeForm'
import { cloneDeep } from 'lodash'
import { newIngredientItem } from '~/composables/type'

interface Validatable {
  validate: () => boolean
}

export default defineComponent({
  name: 'EditRecipeDialog',
  components: {
    RecipeForm
  },
  props: {
    value: {
      type: Boolean,
      default: () => new Recipe()
    },
    recipe: {
      type: Object as PropType<Recipe>,
      default: undefined
    }
  },
  setup (props, { emit }) {
    const recipeBuffer = ref<Recipe>()
    const form = ref<Validatable | null>(null)
    const validates = ref(true)

    // To allow both create and edit
    watch(
      [toRef(props, 'value'), toRef(props, 'recipe')],
      ([value, recipe]) => {
        if (value) {
          if (recipe) {
            recipeBuffer.value = new Recipe(recipe)
          } else {
            recipeBuffer.value = new Recipe({
              ingredients: [newIngredientItem()]
            })
          }
        }
      },
      {
        immediate: true
      }
    )

    const save = () => {
      if (form.value && form.value.validate()) {
        const data = cloneDeep(recipeBuffer.value)
        emit('save', data)
      }
    }

    const close = () => {
      emit('input', false)
    }

    return {
      mdiCancel,
      mdiCheck,
      mdiClose,
      recipeBuffer,
      validates,
      save,
      close
    }
  }
})
</script>
