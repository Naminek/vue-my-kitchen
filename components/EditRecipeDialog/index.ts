import EditRecipeDialog from './EditRecipeDialog.vue'
export default EditRecipeDialog
