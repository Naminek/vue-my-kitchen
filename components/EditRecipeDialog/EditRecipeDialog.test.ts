import Vuetify from 'vuetify'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import EditRecipeDialog from '.'

const localVue = createLocalVue()

describe('EditRecipeDialog', () => {
  const defaultProps = {
    value: true
  }

  beforeEach(() => {
    jest.resetAllMocks()
  })

  function mountFunc (options = {}) {
    return shallowMount(EditRecipeDialog, {
      localVue,
      vuetify: new Vuetify(),
      propsData: defaultProps,
      ...options
    })
  }

  it('should match its snapshot', () => {
    const wrapper = mountFunc()
    expect(wrapper).toMatchSnapshot()
  })
})
