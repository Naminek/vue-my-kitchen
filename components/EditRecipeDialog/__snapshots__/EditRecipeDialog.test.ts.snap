// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditRecipeDialog should match its snapshot 1`] = `
<v-dialog-stub closedelay="0" contentclass maxwidth="1000px" opendelay="0" openonclick="true" origin="center center" persistent="true" retainfocus="true" scrollable="true" transition="dialog-transition" value="true">
  <v-card-stub loaderheight="4" tag="div">
    <v-card-title-stub class="lime lighten-3 white--text font-weight-light text-subtitle-1 text-sm-h6" tag="div">
      Add recipe
      <v-spacer-stub tag="div"></v-spacer-stub>
      <v-btn-stub activeclass class="white--text ml-2" icon="true" tag="button" type="button">
        <v-icon-stub>M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z</v-icon-stub>
      </v-btn-stub>
    </v-card-title-stub>
    <v-card-text-stub class="pt-4" tag="div">
      <v-form-stub lazyvalidation="true" value="true">
        <recipeform-stub recipe="[object Object]"></recipeform-stub>
      </v-form-stub>
    </v-card-text-stub>
    <v-divider-stub></v-divider-stub>
    <v-card-actions-stub tag="div">
      <v-btn-stub activeclass class="mx-2 mx-sm-0" data-cy="saveButton" tag="button" text="true" type="button">
        <v-icon-stub left="true">
          M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z
        </v-icon-stub>
        <span>Save</span>
      </v-btn-stub>
      <v-btn-stub activeclass data-cy="cancelButton" tag="button" text="true" type="button">
        <v-icon-stub>
          M12 2C17.5 2 22 6.5 22 12S17.5 22 12 22 2 17.5 2 12 6.5 2 12 2M12 4C10.1 4 8.4 4.6 7.1 5.7L18.3 16.9C19.3 15.5 20 13.8 20 12C20 7.6 16.4 4 12 4M16.9 18.3L5.7 7.1C4.6 8.4 4 10.1 4 12C4 16.4 7.6 20 12 20C13.9 20 15.6 19.4 16.9 18.3Z
        </v-icon-stub>
        <span>Cancel</span>
      </v-btn-stub>
    </v-card-actions-stub>
  </v-card-stub>
</v-dialog-stub>
`;
